form/form.go

```diff
@@ -338,16 +339,25 @@ func isAppender(t reflect.Type) bool {
 }
 
 func mapEncoder(values *Values, v reflect.Value, keyParts []string, _ bool, _ *formOptions) {
+	keys := make([]string, 0, v.Len())
 	for _, keyVal := range v.MapKeys() {
-		if Strict && keyVal.Kind() != reflect.String {
-			panic("Don't support serializing maps with non-string keys")
+		if keyVal.Kind() != reflect.String {
+			if Strict {
+				panic("Don't support serializing maps with non-string keys")
+			} else {
+				// otherwise keyVal.String() will panic later
+				continue
+			}
 		}
-
+		keys = append(keys, keyVal.String())
+	}
+	sort.Strings(keys)
+	for _, key := range keys {
 		// Unlike a property on a struct which will contain a zero value even
 		// if never set, any value found in a map has been explicitly set, so
 		// we always make an effort to encode them, even if a zero value
 		// (that's why we pass through `true` here).
-		reflectValue(values, v.MapIndex(keyVal), true, append(keyParts, keyVal.String()))
```

Comment:
This was problematic as well because non-string values would always panic, regardless of strict mode. In case of strict mode it would obviously panic with the explicit panic above, but in the non-strict mode it would still panic as `keyVal.String()` would panic.
I added a test for ensuring non-string keys are just ignored.

Reply:
Hm, I tried this test on master and it didn't panic:

```go

func TestEncodeMapNonStringKey(t *testing.T) {
	// Disable strict mode for this test, so the non-string key is ignored.
	Strict = false
	defer func() {
		Strict = true
	}()

	form := &Values{}
	assert.NotPanics(t, func() {
		AppendTo(form, map[int]string{1: "foo"})
	})
	assert.Equal(t, "%3Cint+Value%3E=foo", form.Encode())
}

```

Granted the result is not as useful but we neither had to ignore or panic. What panics in you version is `v.MapIndex(reflect.ValueOf(key))` because the string key is being provided to an int-based map.

Reply:
Ah sorry, you are correct.

> ```go
> // String is a special case because of Go's String method convention.
> // Unlike the other getters, it does not panic if v's Kind is not String.
> ```

I don't know what behavior you prefer at this point, happy to adjust for either.